packages/neptune-extension/src/common/components/upload-modal/UploadForm.tsx

```diff
@@ -0,0 +1,185 @@
+import React from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+
+import { NotebookDTO } from 'generated/leaderboard-client/src/models';
+
+import {
+  PlatformNotebook,
+  PlatformNotebookMetadata,
+} from 'types/platform';
+
+import {
+  uploadNotebook,
+  uploadCheckpoint,
+  getNotebook,
+} from 'common/utils/upload';
+
+import { PROJECT_LOCAL_STORAGE_KEY } from 'common/utils/localStorage';
+
+import Modal from 'common/components/modal/Modal';
+import ProjectInput from 'common/components/input/ProjectInput';
+
+import { getConfigurationState } from 'common/state/configuration/selectors';
+import { getNotebookState } from 'common/state/notebook/selectors'
+import { setNotebook } from 'common/state/notebook/actions'
+
+interface UploadFormProps {
+  platformNotebook: PlatformNotebook
+  onClose: () => void
+}
+
+type UploadMode = 'notebook' | 'checkpoint'
+
+const UploadForm: React.FC<UploadFormProps> = ({
+  platformNotebook,
+  onClose,
+}) => {
+  const [ metadata ] = React.useState(() => platformNotebook.getMetadata());
+
+  const dispatch = useDispatch()
+  const { notebook } = useSelector(getNotebookState)
+  const { inferredUsername } = useSelector(getConfigurationState)
+
+  const noAccess = !notebook && metadata.notebookId
+  const ownerChanged = notebook && notebook.owner !== inferredUsername
+  const pathChanged = notebook && notebook.path !== metadata.path
+
+  const canUploadCheckpoint = notebook && !ownerChanged
+
+  const [ loading, setLoading ] = React.useState(false);
+
+  const [ mode, setMode ] = React.useState<UploadMode>(canUploadCheckpoint ? 'checkpoint' : 'notebook');
+
+  const [ projectId, setProjectId ] = React.useState(() => {
+    return !!notebook
+      ? notebook.projectId
+      : window.localStorage.getItem(PROJECT_LOCAL_STORAGE_KEY) || ''
+  });
+  const [ name, setName ] = React.useState('');
+  const [ description, setDescription ] = React.useState('');
+
+  function changeMode(newMode: UploadMode) {
+    if (newMode === 'checkpoint') {
+      setProjectId((notebook as NotebookDTO).projectId);
+    }
+
+    setMode(newMode);
+  }
+
+  async function handleSubmit() {
+    setLoading(true);
+
+    const content = await platformNotebook.getContent();
+
+    const checkpointMeta = {
+      path: metadata.path,
+      name,
+      description,
+    };
+
+    if (mode === 'notebook') {
+      const notebookMeta = await uploadNotebook(projectId, checkpointMeta, content);
+
+      await platformNotebook.saveNotebookId(notebookMeta.id);
+
+      const notebook = await getNotebook(notebookMeta.id)
+      dispatch(setNotebook(notebook))
+    }
+    else {
+      await uploadCheckpoint(metadata.notebookId as string, checkpointMeta, content);
```

Comment:
This casting is not needed and potentially it hides an error.
If we don't have metadata.notebookId, we shouldnt call uploadCheckpoint

Reply:
The logic you have just descibed is evident from the rest of code, so this is just a way of saying "I am sure this will not be undefined". Typescript type inference does not go that far so on other hand you would just have to introduce type checks that would be dead code anyway. I am not sure which technique is better and have mixed fellings about it.

Reply:
I see, I would at least add comment here?

Reply:
This typescript declaration serves as a comment because you make it explicit in that point that a value is a string. With plain js you do not have such syntax. It only requires a mutual agreement that such statements are concious writing and not just a way of avoiding compiler complaints, but this goes without saying I guess. 

But I can write a word or two, just to explain why this is string (because you cannot upload, if a notebook id was not created).